src/dos/cdrom_image.cpp

```diff
@@ -70,80 +89,155 @@ int CDROM_Interface_Image::BinaryFile::getLength()
+#endif
+}
+
+
+bool CDROM_Interface_Image::BinaryFile::seek(Bit32u offset)
+{
+	file->seekg(offset, ios::beg);
+	return !file->fail();
+}
+
+Bit16u CDROM_Interface_Image::BinaryFile::decode(Bit8u *buffer)
+{
+	file->read((char*)buffer, chunkSize);
+	return (Bit16u) file->gcount();
+}
+
 CDROM_Interface_Image::AudioFile::AudioFile(const char *filename, bool &error)
+	:TrackFile(4096)
 {
-	Sound_AudioInfo desired = {AUDIO_S16, 2, 44100};
-	sample = Sound_NewSampleFromFile(filename, &desired, RAW_SECTOR_SIZE);
-	lastCount = RAW_SECTOR_SIZE;
-	lastSeek = 0;
-	error = (sample == NULL);
+	// Use the audio file's actual sample rate and number of channels as opposed to overriding
+	Sound_AudioInfo desired = {AUDIO_S16, 0, 0};
+	sample = Sound_NewSampleFromFile(filename, &desired, chunkSize);
+	if (sample) {
+		error = false;
+		std::string filename_only(filename);
+		filename_only = filename_only.substr(filename_only.find_last_of("\\/") + 1);
+		LOG_MSG("CDROM: Loaded %s [%d Hz %d-channel]", filename_only.c_str(), this->getRate(), this->getChannels());
+	} else
+		error = true;
 }
 
 CDROM_Interface_Image::AudioFile::~AudioFile()
 {
 	Sound_FreeSample(sample);
 }
 
-bool CDROM_Interface_Image::AudioFile::read(Bit8u *buffer, int seek, int count)
+bool CDROM_Interface_Image::AudioFile::seek(Bit32u offset)
 {
-	if (lastCount != count) {
-		int success = Sound_SetBufferSize(sample, count);
-		if (!success) return false;
-	}
-	if (lastSeek != (seek - count)) {
-		int success = Sound_Seek(sample, (int)((double)(seek) / 176.4f));
-		if (!success) return false;
+/*
+// The following code requires C++11 otherwise creating an ANSI sub-second timer
+// that's portable across Windows, Linux, and macOS requires a very lengthly solution:
+// https://stackoverflow.com/questions/361363/how-to-measure-time-in-milliseconds-using-ansi-c
+// Leave this in place (but commented out) for development and debugging purposes.
+//
+#ifdef DEBUG
+	const auto begin = std::chrono::steady_clock::now();
+#endif
+*/
+	// Convert the byte-offset to a time offset (milliseconds)
+	const bool result = Sound_Seek(sample, lround(offset/176.4f));
+
+/*
+#ifdef DEBUG
+	const auto end = std::chrono::steady_clock::now();
+	LOG_MSG("%s CDROM: seek(%u) took %f ms", get_time(), offset, chrono::duration <double, milli> (end - begin).count());
+#endif
+*/
+	return result;
+}
+
+Bit16u CDROM_Interface_Image::AudioFile::decode(Bit8u *buffer)
+{
+	const Bit16u bytes = Sound_Decode(sample);
+	memcpy(buffer, sample->buffer, bytes);
+	return bytes;
+}
+
+Bit16u CDROM_Interface_Image::AudioFile::getEndian()
+{
+	return sample->actual.format;
+}
+
+Bit32u CDROM_Interface_Image::AudioFile::getRate()
+{
+	Bit32u rate(0);
+	if (sample) {
+		rate = sample->actual.rate;
 	}
-	lastSeek = seek;
-	int bytes = Sound_Decode(sample);
-	if (bytes < count) {
-		memcpy(buffer, sample->buffer, bytes);
-		memset(buffer + bytes, 0, count - bytes);
-	} else {
-		memcpy(buffer, sample->buffer, count);
+	return rate;
```

Comment:
return sample ? sample->actual.rate : 0;

?

Reply:
This is very nice! Fixed in 4e96f038